crates/core/src/protocols/fri/common.rs

```diff
@@ -38,66 +36,69 @@ where
 /// precondition checks.
 ///
-/// See [DP24], Def. 3.6 and Lemma 3.9 for more details.
+/// ## Arguments
+///
+/// * `ntt` - the NTT instance, used to look up the twiddle values.
+/// * `log_len` - the binary logarithm of the code length.
+/// * `chunk_index` - the index of the chunk, of size $2^n$, in the full codeword.
+/// * `values` - mutable slice of values to fold, modified in place.
+/// * `challenges` - the sequence of folding challenges, with length $n$.
+///
+/// ## Pre-conditions
+///
+/// - `challenges.len() <= log_len`.
+/// - `log_len <= ntt.log_domain_size()`, so that the NTT domain is large enough.
+/// - `values.len() == 1 << challenges.len()`.
 ///
 /// [DP24]: <https://eprint.iacr.org/2024/504>
 #[inline]
 pub fn fold_chunk<F, FS, NTT>(
 	ntt: &NTT,
-	start_round: usize,
+	mut log_len: usize,
 	chunk_index: usize,
-	values: &[F],
-	folding_challenges: &[F],
-	scratch_buffer: &mut [F],
+	values: &mut [F],
+	challenges: &[F],
 ) -> F
 where
 	F: BinaryField + ExtensionField<FS>,
 	FS: BinaryField,
 	NTT: AdditiveNTT<FS>,
 {
+	let mut log_size = challenges.len();
+
 	// Preconditions
-	debug_assert!(!folding_challenges.is_empty());
-	debug_assert!(start_round + folding_challenges.len() <= ntt.log_domain_size());
-	debug_assert_eq!(values.len(), 1 << folding_challenges.len());
-	debug_assert!(scratch_buffer.len() >= values.len());
-
-	// Fold the chunk with the folding challenges one by one
-	for n_challenges_processed in 0..folding_challenges.len() {
-		let n_remaining_challenges = folding_challenges.len() - n_challenges_processed;
-		let scratch_buffer_len = values.len() >> n_challenges_processed;
-		let new_scratch_buffer_len = scratch_buffer_len >> 1;
-		let round = start_round + n_challenges_processed;
-		let r = folding_challenges[n_challenges_processed];
-		let index_start = chunk_index << (n_remaining_challenges - 1);
+	debug_assert!(log_size <= log_len);
+	debug_assert!(log_len <= ntt.log_domain_size());
+	debug_assert_eq!(values.len(), 1 << log_size);
 
+	// FRI-fold the values in place.
+	for &challenge in challenges {
 		// Fold the (2i) and (2i+1)th cells of the scratch buffer in-place into the i-th cell
-		if n_challenges_processed > 0 {
-			(0..new_scratch_buffer_len).for_each(|index_offset| {
-				let values =
-					(scratch_buffer[index_offset << 1], scratch_buffer[(index_offset << 1) + 1]);
-				scratch_buffer[index_offset] =
-					fold_pair(ntt, round, index_start + index_offset, values, r)
-			});
-		} else {
-			// For the first round, we read values directly from the `values` slice.
-			(0..new_scratch_buffer_len).for_each(|index_offset| {
-				let values = (values[index_offset << 1], values[(index_offset << 1) + 1]);
-				scratch_buffer[index_offset] =
-					fold_pair(ntt, round, index_start + index_offset, values, r)
-			});
+		let ntt_round = ntt.log_domain_size() - log_len;
+		for index_offset in 0..1 << (log_size - 1) {
+			let pair = (values[index_offset << 1], values[(index_offset << 1) + 1]);
```

Comment:
```suggestion
			let pair = (values[index_offset << 1], values[index_offset << 1 | 1]);
```
not sure if it matters? but in general bitwise OR prob more efficient than add?

Reply:
Not on CPUs. And our linter requires parenthesis here :-)